frame/support/src/traits.rs

```diff
@@ -1518,6 +1518,26 @@ pub trait OnFinalize<BlockNumber> {
 	fn on_finalize(_n: BlockNumber) {}
 }
 
+pub trait OnIdle<BlockNumber> {
+	/// The block is being finalized. Implement to have something happen.
+	///
+	/// NOTE: This function is called AFTER ALL extrinsics in a block are applied,
+	/// including inherent extrinsics.
+	fn on_idle(_n: BlockNumber, _remaining_weight: crate::weights::Weight) -> crate::weights::Weight { 0 }
+}
+
+#[impl_for_tuples(30)]
+impl<BlockNumber: Clone> OnIdle<BlockNumber> for Tuple {
+	fn on_idle(n: BlockNumber,  remaining_weight: crate::weights::Weight) -> crate::weights::Weight {
+		let mut weight = 0;
+		for_tuples!( #(
+			let adjusted_remaining_weight = remaining_weight.saturating_sub(weight);
+			weight = weight.saturating_add(Tuple::on_idle(n.clone(),  adjusted_remaining_weight));
```

Comment:
I wonder if we should do `register_extra_weight_unchecked` here after each call.
I think it is fine we probably don't need to do.

Reply:
We could return as soon as the weight reaches zero, instead. No need to call the hooks with zero weight available, right?

Reply:
Yes we can also do that, but my point was more. if a function inside on_idle read the value of BlockWeight storage, the storage will not have the updated value. The value is only updated at the end of all on_idle executions.

Reply:
You are right. But should they do that? Thos hooks should respect the weight that is passed to them.

Reply:
Yes indeed they shouldn't

Reply:
mmh, but if they execute arbitrary code that could include checking the weight.
If we imagine the scheduler executing something or some migration running I could see this happening. Would be quite the hellish bug to track down

Reply:
I see no harm on calling `register_extra_weight_unchecked` other than that it adds a lot of host function calls which is a potential performance hazard.

If we go down that route we should definitely break out of the loop once we consumed all weight.

Reply:
@JesseAbram just figured out that we cannot do this in `frame_support` because it would lead to a cyclic dependency :-/

Reply:
ya, after talking about it with @apopiak I think leaving it as is, and documenting the issue with getting the weight on on_idle from frame_system